dask/async.py

```diff
@@ -458,16 +463,28 @@ def fire_task():
             data = dict((dep, state['cache'][dep])
                         for dep in get_dependencies(dsk, key))
             # Submit
-            apply_async(execute_task, args=[key, dsk[key], data, queue,
-                                            get_id, raise_on_exception])
+            apply_async(execute_task,
+                        args=(key, marshall((dsk[key], data)),
+                              marshall, unmarshall,
+                              get_id, raise_on_exception),
+                        callback=queue.put)
+
+        def fill_pool_with_ready_tasks():
+            while state['ready'] and len(state['running']) < num_workers:
+                fire_task()
 
         # Seed initial tasks into the thread pool
-        while state['ready'] and len(state['running']) < num_workers:
-            fire_task()
+        fill_pool_with_ready_tasks()
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res, tb, worker_id = queue.get()
+            key, res_info = queue.get()
+            try:
+                res, tb, worker_id = unmarshall(res_info)
+            except Exception as e:
```

Comment:
Can you show me where in the code this deserialization error catching (in the main process, not in the workers) was implemented? This appears to be new functionality to me.

The point of the whole `remote_exception` was to indicate that the exception occurred in a separate process, and display the original exception and traceback in a way most useful to the user. The deserialization exceptions fail in the main process, and the traceback from the deserialization failure should be easy to read already, as it's in the main process.

The way it's implemented in this PR, we catch a deserialization error, grab the current traceback, cleanup the scheduler state, then raise a remote_exception with the deserialization error and the traceback from that error. This has nothing to do with any remote errors, and could occur if the result of the task was good (but fails to deserialize) or bad (and the exception fails to deserialize).